parser/parser.go

```diff
@@ -158,30 +158,41 @@ func (re *recursionError) Error() string {
 var _ error = &recursionError{}
 
 type recursionListener struct {
-	maxDepth int
-	depth    int
+	maxDepth      int
+	ruleTypeDepth map[int]int
 }
 
 func (rl *recursionListener) VisitTerminal(node antlr.TerminalNode) {}
 
 func (rl *recursionListener) VisitErrorNode(node antlr.ErrorNode) {}
 
 func (rl *recursionListener) EnterEveryRule(ctx antlr.ParserRuleContext) {
-	if ctx != nil && ctx.GetRuleIndex() == gen.CELParserRULE_expr {
-		if rl.depth >= rl.maxDepth {
-			rl.depth++
-			panic(&recursionError{
-				message: fmt.Sprintf("expression recursion limit exceeded: %d", rl.maxDepth),
-			})
-		}
-		rl.depth++
+	if ctx == nil {
+		return
+	}
+	ruleIndex := ctx.GetRuleIndex()
```

Comment:
I *think* this is safe. I remember having to put the increment after the recursion check. But looking at ANTLR Go, it appears that if EnterEveryRule panics, ExitEveryRule should not be called. Might want to double-check, otherwise things could go negative.

Reply:
Done. Added a defensive check since I'm using a pointer now as well.